cmd/retry.go

```diff
@@ -0,0 +1,63 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	json "github.com/minio/colorjson"
+	"github.com/minio/mc/pkg/probe"
+)
+
+type retryManager struct {
+	retries       int
+	maxRetries    int
+	retryInterval time.Duration
+	shouldStop    bool
+	ctx           context.Context
+}
+
+func newRetryManager(ctx context.Context, retryInterval time.Duration, maxRetries int) *retryManager {
+	return &retryManager{
+		retryInterval: retryInterval,
+		maxRetries:    maxRetries,
+		ctx:           ctx,
+	}
+}
+
+type retryMessage struct {
+	SourceURL string `json:"sourceURL"`
+	TargetURL string `json:"targetURL"`
+	Retries   int    `json:"retries"`
+}
+
+func (r *retryManager) Stop() {
+	r.shouldStop = true
+}
+
+func (r retryMessage) String() string {
+	return fmt.Sprintf("<INFO> Retries %d: source `%s` >> target `%s`", r.Retries, r.SourceURL, r.TargetURL)
+}
+
+func (r retryMessage) JSON() string {
+	jsonMessageBytes, e := json.MarshalIndent(r, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(jsonMessageBytes)
+}
+
+func (r *retryManager) encapsulateWithRetry(action func(*retryManager) *probe.Error) {
```

Comment:
```suggestion
func (r *retryManager) retry(action func(*retryManager) *probe.Error) {
```
If you are using context for cancelation:

```suggestion
func (r *retryManager) retry(action func(context.Context, *retryManager) *probe.Error) {
```